### PR TITLE
feat: [DEL-3663]: Add mTLS support to delegate code.

### PR DESCRIPTION
**Pull request by johannesHarness**
## Describe your changes
This change is adding mTLS functionality to delegate agent code.
For more details on mTLS delegate design, please see [mTLS Delegate - Design Document](https://harness.atlassian.net/wiki/spaces/~203497205/pages/21043576978/mTLS+Delegate+-+Design+Document)

**IMPORTANT** Functionality is added to immutable delegate only.

This PR is adding 4 new fields to the delegate config:
- `clientCertificateFilePath` + `clientCertificateKeyFilePath`
required to pass the path to the certificate file and key used for mTLS during TLS handshake
- `grpcAuthorityModificationDisabled`:
required to use original FQDN as authority for GRPC calls since the delegate-gateway uses the metadata for routing
- `trustAllCertificates`:
required for local testing with self signed certificate, if true all certs are trusted (never set in generated yaml)

To allow for easier creation of TrustManagers and KeyManagers and unify the usage across the different clients (not everywhere, but at least across all delegate related code), I've added the `X509TrustManagerBuilder`, `X509KeyManagerBuilder` and `X509SslContextBuilder`.

### Testing
I've tested mTLS both with local running Intelij delegate against PR environment and immutable delegate deployed in PR environment. Results can be seen here:

https://harness.atlassian.net/wiki/spaces/~203497205/pages/21102100675/Delegate+Changes

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/34139)

### CODE REVIEW

**Comment by johannesHarness** [comment]
trigger publish-delegate

**Comment by johannesHarness** [comment]
trigger publish-delegate

**Comment by johannesHarness** [comment]
trigger publish-delegate

**Comment by johannesHarness** [comment]
trigger publish-delegate

**Comment by johannesHarness** [comment]
trigger publish-delegate

**Comment by johannesHarness** [comment]
trigger publish-delegate

File: 960-watcher/src/main/java/io/harness/managerclient/WatcherManagerClientV2Factory.java [hunk 1]
```diff
@@ -59,16 +55,18 @@ public ManagerClientV2 get() {
   private OkHttpClient getUnsafeOkHttpClient() {
     try {
       // Install the all-trusting trust manager
+      X509TrustManager[] trustManagers =
+          new X509TrustManager[] {new X509TrustManagerBuilder().trustAllCertificates().build()};
       final SSLContext sslContext = SSLContext.getInstance("SSL");
-      sslContext.init(null, TRUST_ALL_CERTS.toArray(new TrustManager[1]), new java.security.SecureRandom());
+      sslContext.init(null, trustManagers, new java.security.SecureRandom());
```
Comment (johannesHarness) [inline 960-watcher/src/main/java/io/harness/managerclient/WatcherManagerClientV2Factory.java:61]: change to use the TRUST_ALL_CERTS but with new allTrustingTrustmanager

File: 980-commons/src/main/java/io/harness/security/AllTrustingX509TrustManager.java [hunk 2]
```diff
@@ -1,16 +1,19 @@
 /*
- * Copyright 2021 Harness Inc. All rights reserved.
```
Comment (johannesHarness) [inline 980-commons/src/main/java/io/harness/security/AllTrustingX509TrustManager.java:2]: change back to original

**Comment by raghuAtWings** [comment]
*[`260-delegate/src/main/java/io/harness/delegate/app/DelegateAgentConfig.java` line 42 at r4](https://reviewable.io/reviews/harness/harness-core/34139#-N529q8G4_kxmV8IM0XF:-N529q8HBETyXivvYrkB:b2ryge5) ([raw file](https://github.com/harness/harness-core/blob/21c617491dfb1a207d92f9f4dbe869fca2cf6061/260-delegate/src/main/java/io/harness/delegate/app/DelegateAgentConfig.java#L42)):*
> ```Java
>   @JsonProperty private final String clientCertificateKeyFilePath;
>   @JsonProperty private final boolean trustAllCertificates;
>   @JsonProperty private final boolean grpcAuthorityModificationDisabled;
> ```

we are getting rid of grpc calls from delegate to any microservice, we can get rid of grpc related changes

**Comment by raghuAtWings** [comment]
*[`270-verification/src/main/java/io/harness/managerclient/VerificationManagerClientFactory.java` line 35 at r5](https://reviewable.io/reviews/harness/harness-core/34139#-N5a5zct6anjnBSLqDIX:-N5a5zct6anjnBSLqDIY:b-2njekr) ([raw file](https://github.com/harness/harness-core/blob/14350b6477ec555109429afaf3f1f051b8089076/270-verification/src/main/java/io/harness/managerclient/VerificationManagerClientFactory.java#L35)):*
> ```Java
> public class VerificationManagerClientFactory implements Provider<VerificationManagerClient> {
>   public static final ImmutableList<TrustManager> TRUST_ALL_CERTS =
>       ImmutableList.of(new VerificationManagerClientX509TrustManager());
> ```

not following this change? Does it mean verification client will always trust?

**Review by raghuAtWings — COMMENTED** [review]
> Reviewed 22 of 41 files at r1, 5 of 6 files at r2, 1 of 1 files at r3, 18 of 20 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34139)* status: Some of the checks are too old. 0 of 1 approvals obtained.

**Comment by raghuAtWings** [comment]
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 750 at r5](https://reviewable.io/reviews/harness/harness-core/34139#-N5aF9Kf0W-LPeAWJoTl:-N5aF9Kf0W-LPeAWJoTm:b-xmndpf) ([raw file](https://github.com/harness/harness-core/blob/a93e2c38bde9881c9241c213639c0f9c8c5c31f3/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L750)):*
> ```Java
> 
>       // send accountId + delegateId as header for delegate gateway to log websocket connection with account.
>       requestBuilder.header("accountId", this.delegateConfiguration.getAccountId());
> ```

see if this can also be done with this PR

https://harness.atlassian.net/browse/DEL-4183

**Review by raghuAtWings — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34139)* status: Some of the checks are too old.

**Comment by johannesHarness** [comment]
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 750 at r5](https://reviewable.io/reviews/harness/harness-core/34139#-N5aF9Kf0W-LPeAWJoTl:-N5d4DNn8mrpGlbcgMLi:b-yfpq50) ([raw file](https://github.com/harness/harness-core/blob/a93e2c38bde9881c9241c213639c0f9c8c5c31f3/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L750)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

see if this can also be done with this PR

https://harness.atlassian.net/browse/DEL-4183
</blockquote></details>

I'd prefer to not further delay this PR and do it in a separate PR, as I don't see any benefit of having it part of it in this one.

**Review by mdmilic — CHANGES_REQUESTED** [review]
> Reviewed 22 of 41 files at r1, 5 of 6 files at r2, 1 of 1 files at r3, 16 of 20 files at r4, 1 of 3 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34139)* status: 1 change requests. 0 of 1 approvals obtained. 

___
*[`950-log-client/src/main/java/io/harness/logstreaming/LogStreamingClientFactory.java` line 77 at r6](https://reviewable.io/reviews/harness/harness-core/34139#-N5f9vFDEfisuXLhunZg:-N5f9vFDEfisuXLhunZh:b-cmk7ud) ([raw file](https://github.com/harness/harness-core/blob/366219ff9edde2deed772b323dc9080f26873a98/950-log-client/src/main/java/io/harness/logstreaming/LogStreamingClientFactory.java#L77)):*
> ```Java
>   }
> 
>   private OkHttpClient getSafeOkHttpClient() {
> ```

All these client factories look pretty much the same, maybe we should extract everything useful to a common class and extend it. 
___
*[`970-grpc/src/test/java/io/harness/grpc/utils/DelegateGrpcConfigExtractorTest.java` line 58 at r6](https://reviewable.io/reviews/harness/harness-core/34139#-N5fBAMDBC5x7yMIiJb2:-N5fBAMDBC5x7yMIiJb3:bj5u8tv) ([raw file](https://github.com/harness/harness-core/blob/366219ff9edde2deed772b323dc9080f26873a98/970-grpc/src/test/java/io/harness/grpc/utils/DelegateGrpcConfigExtractorTest.java#L58)):*
> ```Java
> 
>   @Test
>   @Owner(developers = AVMOHAN)
> ```

Pretending to be someone else :smiley: 
___
*[`980-commons/src/main/java/io/harness/security/X509KeyManagerBuilder.java` line 50 at r6](https://reviewable.io/reviews/harness/harness-core/34139#-N5fDvH53I22OdQJtwn9:-N5fDvH53I22OdQJtwnA:b-uuieo3) ([raw file](https://github.com/harness/harness-core/blob/366219ff9edde2deed772b323dc9080f26873a98/980-commons/src/main/java/io/harness/security/X509KeyManagerBuilder.java#L50)):*
> ```Java
>   private static final String KEY_MANAGER_ALGORITHM_SUNX509 = "SunX509";
>   private static final String KEY_ALGORITHM_RSA = "RSA";
>   private static final String KEY_STORE_TYPE_JAVA = "jks";
> ```

Please double check the defaults now after JDK11 migration. I know PKCS12 is now default keystore (and is more secure) though JKS would still work. Not sure about other. params. 
___
*[`980-commons/src/main/java/io/harness/security/X509KeyManagerBuilder.java` line 76 at r6](https://reviewable.io/reviews/harness/harness-core/34139#-N5fHM2v1p3N68fpg1Y_:-N5fHM2w7r6wqyJSmAiF:bud2rsu) ([raw file](https://github.com/harness/harness-core/blob/366219ff9edde2deed772b323dc9080f26873a98/980-commons/src/main/java/io/harness/security/X509KeyManagerBuilder.java#L76)):*
> ```Java
>     try {
>       this.keystore = KeyStore.getInstance(KEY_STORE_TYPE_JAVA);
>       this.keystore.load(null, null);
> ```

Would all of these work if customers provide their own custom keystore (don't know if Java supports multiple keystores or one overrides the other). 
Additionally how about asking customers to store the certs in the custom keystore and just have them provide that rather than raw cert files (we already have this for some scenarios like 3rd party tool certs or proxy certs etc...)
Check this article out
https://docs.harness.io/article/lb2cxxgak1-trust-store-override-for-delegates

**Review by johannesHarness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34139)* status: 7 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/delegate/app/DelegateAgentConfig.java` line 42 at r4](https://reviewable.io/reviews/harness/harness-core/34139#-N529q8G4_kxmV8IM0XF:-N5iFfzy6lOZCWxCKjH4:b-7g7q9t) ([raw file](https://github.com/harness/harness-core/blob/21c617491dfb1a207d92f9f4dbe869fca2cf6061/260-delegate/src/main/java/io/harness/delegate/app/DelegateAgentConfig.java#L42)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

we are getting rid of grpc calls from delegate to any microservice, we can get rid of grpc related changes 
</blockquote></details>

As discussed, keeping it to avoid being blocked on grpc work.
___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 750 at r5](https://reviewable.io/reviews/harness/harness-core/34139#-N5a4Ows1JcP-1K-oNDK:-N5iFob4AWxGypfZPxKP:b-g5b1hz) ([raw file](https://github.com/harness/harness-core/blob/a93e2c38bde9881c9241c213639c0f9c8c5c31f3/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L750)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

see if this can also be done with this PR

https://harness.atlassian.net/browse/DEL-4183
</blockquote></details>

I think that work should be separated into its own PR.
___
*[`270-verification/src/main/java/io/harness/managerclient/VerificationManagerClientFactory.java` line 35 at r5](https://reviewable.io/reviews/harness/harness-core/34139#-N5a5zct6anjnBSLqDIX:-N5iFvgJ2Tqfokc3IPRv:b-trzgyo) ([raw file](https://github.com/harness/harness-core/blob/14350b6477ec555109429afaf3f1f051b8089076/270-verification/src/main/java/io/harness/managerclient/VerificationManagerClientFactory.java#L35)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

not following this change? Does it mean verification client will always trust?
</blockquote></details>

As discussed, will keep trust behavior as it was before the change. Will have to change that in the future.
___
*[`950-log-client/src/main/java/io/harness/logstreaming/LogStreamingClientFactory.java` line 77 at r6](https://reviewable.io/reviews/harness/harness-core/34139#-N5f9vFDEfisuXLhunZg:-N5k_NvV-KSVmuWMPPMu:bhsa89z) ([raw file](https://github.com/harness/harness-core/blob/366219ff9edde2deed772b323dc9080f26873a98/950-log-client/src/main/java/io/harness/logstreaming/LogStreamingClientFactory.java#L77)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

All these client factories look pretty much the same, maybe we should extract everything useful to a common class and extend it. 
</blockquote></details>

As discussed, will consider in future PR.
___
*[`970-grpc/src/test/java/io/harness/grpc/utils/DelegateGrpcConfigExtractorTest.java` line 58 at r6](https://reviewable.io/reviews/harness/harness-core/34139#-N5fBAMDBC5x7yMIiJb2:-N5iG8R18uMJLq0yLU9D:b-ve3yvb) ([raw file](https://github.com/harness/harness-core/blob/366219ff9edde2deed772b323dc9080f26873a98/970-grpc/src/test/java/io/harness/grpc/utils/DelegateGrpcConfigExtractorTest.java#L58)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

Pretending to be someone else :smiley: 
</blockquote></details>

:smile: 
___
*[`980-commons/src/main/java/io/harness/security/X509KeyManagerBuilder.java` line 50 at r6](https://reviewable.io/reviews/harness/harness-core/34139#-N5fDvH53I22OdQJtwn9:-N5k_FevAsv1rLmpaWTg:bm0edeu) ([raw file](https://github.com/harness/harness-core/blob/366219ff9edde2deed772b323dc9080f26873a98/980-commons/src/main/java/io/harness/security/X509KeyManagerBuilder.java#L50)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

Please double check the defaults now after JDK11 migration. I know PKCS12 is now default keystore (and is more secure) though JKS would still work. Not sure about other. params. 
</blockquote></details>

I changed to use the default algo (though that's for internally generated keystore, not passed by customer)
___
*[`980-commons/src/main/java/io/harness/security/X509KeyManagerBuilder.java` line 76 at r6](https://reviewable.io/reviews/harness/harness-core/34139#-N5fHM2v1p3N68fpg1Y_:-N5iGNMC7IORq-0g11Wr:b-v5z58i) ([raw file](https://github.com/harness/harness-core/blob/366219ff9edde2deed772b323dc9080f26873a98/980-commons/src/main/java/io/harness/security/X509KeyManagerBuilder.java#L76)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

Would all of these work if customers provide their own custom keystore (don't know if Java supports multiple keystores or one overrides the other). 
Additionally how about asking customers to store the certs in the custom keystore and just have them provide that rather than raw cert files (we already have this for some scenarios like 3rd party tool certs or proxy certs etc...)
Check this article out
https://docs.harness.io/article/lb2cxxgak1-trust-store-override-for-delegates
</blockquote></details>

Talked with Puneet, will keep it as is to continue using widely used PEM format instead of pkcs12.

**Review by johannesHarness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34139)* status: 7 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`980-commons/src/main/java/io/harness/security/X509KeyManagerBuilder.java` line 76 at r6](https://reviewable.io/reviews/harness/harness-core/34139#-N5fHM2v1p3N68fpg1Y_:-N5kaMRE4muF-S_WDYlQ:bt7s0mv) ([raw file](https://github.com/harness/harness-core/blob/366219ff9edde2deed772b323dc9080f26873a98/980-commons/src/main/java/io/harness/security/X509KeyManagerBuilder.java#L76)):*
<details><summary><i>Previously, johannesHarness (Johannes Batzill) wrote…</i></summary><blockquote>

Talked with Puneet, will keep it as is to continue using widely used PEM format instead of pkcs12.
</blockquote></details>

As for the custom keystore, I tested overwriting the system settings locally and it loaded the custom cert instead of the default keystore (-Djavax.net.ssl.trustStore=/tmp/agent-gateway/certs/server/keystore.p12 -Djavax.net.ssl.trustStoreType=pkcs12 -Djavax.net.ssl.trustStorePassword=changeit").

Thanks for bringing that up, good that I tested it!

**Comment by johannesHarness** [comment]
trigger ti3

**Review by mdmilic — APPROVED** [review]
> Reviewed 7 of 7 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34139)* status: :shipit: complete! ✓

**Comment by johannesHarness** [comment]
trigger ti0

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 1 of 2 files at r6, 7 of 7 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34139)* status: :shipit: complete! ✓

**Review by Vbhatia20 — APPROVED** [review]
> Reviewed 21 of 41 files at r1, 3 of 6 files at r2, 1 of 1 files at r3, 13 of 20 files at r4, 1 of 3 files at r5, 6 of 7 files at r7, 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34139)* status: :shipit: complete! ✓

**Review by arpit3107 — APPROVED** [review]
> Reviewed 21 of 41 files at r1, 3 of 6 files at r2, 1 of 1 files at r3, 13 of 20 files at r4, 1 of 3 files at r5, 6 of 7 files at r7, 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34139)* status: :shipit: complete! ✓

**Comment by johannesHarness** [comment]
*[`420-delegate-agent/src/main/java/io/harness/managerclient/DelegateAgentManagerClientFactory.java` line 83 at r5](https://reviewable.io/reviews/harness/harness-core/34139#-N5a73FF-iAX-cDLbw0i:-N5oG2nO09EH1aKGgyMD:bafkmy8) ([raw file](https://github.com/harness/harness-core/blob/a93e2c38bde9881c9241c213639c0f9c8c5c31f3/420-delegate-agent/src/main/java/io/harness/managerclient/DelegateAgentManagerClientFactory.java#L83)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

not following this change. what does trustDefaultTrustStore() do?
</blockquote></details>

for future reference: it loads the certs from the defaulttruststore (I kept implementation as it was before - without the local cert that was loaded for testing purposes)

**Review by kapil-harness — APPROVED** [review]
> Reviewed 21 of 41 files at r1, 3 of 6 files at r2, 1 of 1 files at r3, 13 of 20 files at r4, 1 of 3 files at r5, 6 of 7 files at r7, 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34139)* status: :shipit: complete! ✓

**Pull request merged**